File: java/1.java
// Two sum

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

class Solution {
  public int[] twoSum(int[] nums, int target) {
    int index = 0;
    HashMap<Integer, Integer> numsH = new HashMap<Integer, Integer>();
    List<Integer> list = Arrays.stream(nums).boxed().collect(Collectors.toList());

    if ((nums[0] + nums[nums.length - 1]) == target) {
      return new int[] {0, nums.length - 1};
    }

    if ((nums[nums.length - 2] + nums[nums.length - 1]) == target) {
      return new int[] {nums.length - 2, nums.length - 1};
    }

    for (int setNums : nums) {
      int ans = target - setNums;

      if (numsH.containsValue(ans)) {
        return new int[] {list.indexOf(ans), index};
      }

      numsH.put(index, setNums);

      ++index;
    }

    return nums;
  }
}

File: cpp/1.cpp
// https://leetcode.com/problems/two-sum

#include <array>
#include <iostream>
#include <unordered_map>
#include <vector>

template<typename M, typename K>
bool contains(const M &m, const K &k) {
    return m.find(k) != m.end();
}

class Solution {
    public:
        std::vector<int> two_sum(const std::vector<int> &nums, const int target)
            const noexcept {
            auto map = std::unordered_map<int, int>{};
            for (auto i = 0ULL, n = nums.size(); i < n; ++i) {
                // TODO
                auto complement = target - nums[i];
                //  map.find(complement) != map.end()
                if (map.contains(complement)) {
                    return {map[complement], static_cast<int>(i)};
                }
                map[nums[i]] = i;
            }
            return nums;
        }
};

int main() {
    auto sol = Solution{};
    for (auto i : sol.two_sum({3, 3}, 6))
        std::cout << i << ' ';
    std::cout << '\n';
    return 0;
}
